src/Microsoft.ML.Data/DataLoadSave/PartitionedFileLoader.cs

```diff
@@ -0,0 +1,726 @@
+            public DataKind? Type;
+
+            [Argument(ArgumentType.Required, HelpText = "Source index of the column.")]
+            public int Source;
+
+            public static Column Parse(string str)
+            {
+                Contracts.AssertNonEmpty(str);
+
+                if (TryParse(str, out Column column))
+                {
+                    return column;
+                }
+
+                return null;
+            }
+
+            public static bool TryParse(string str, out Column column)
+            {
+                column = null;
+
+                if (string.IsNullOrEmpty(str))
+                {
+                    return false;
+                }
+
+                if (!ColumnParsingUtils.TryParse(str, out string name, out string sourceStr, out string kindStr))
+                {
+                    return false;
+                }
+
+                DataKind? kind = null;
+                if (kindStr != null && TypeParsingUtils.TryParseDataKind(kindStr, out DataKind parsedKind, out KeyRange range))
+                {
+                    kind = parsedKind;
+                }
+
+                if (!int.TryParse(sourceStr, out int source))
+                {
+                    return false;
+                }
+
+                column = new Column()
+                {
+                    Name = name,
+                    Source = source,
+                    Type = kind
+                };
+
+                return true;
+            }
+
+            public bool TryUnparse(StringBuilder sb)
+            {
+                Contracts.AssertValue(sb);
+
+                sb.Append($"{Name}");
+
+                if (Type.HasValue)
+                {
+                    sb.Append($":{Type}");
+                }
+
+                sb.Append($":{Source}");
+
+                return true;
+            }
+        }
+
+        private readonly IHost _host;
+        private readonly IMultiStreamSource _files;
+        private readonly Column[] _columns;
+
+        // Number of tailing directories to include.
+        private readonly int _tailingDirCount;
+
+        // An underlying loader used on each individual loader.
+        private readonly SubComponent<IDataLoader, SignatureDataLoader> _subLoader;
+
+        private readonly IPartitionedPathParser _pathParser;
+
+        private const string RegistrationName = LoadName;
+
+        public PartitionedFileLoader(IHostEnvironment env, Arguments args, IMultiStreamSource files)
+        {
+            Contracts.CheckValue(env, nameof(env));
+            _host = env.Register(RegistrationName);
+            _host.CheckValue(args, nameof(args));
+            _host.CheckValue(files, nameof(files));
+
+            _pathParser = args.PathParserFactory.CreateComponent(_host);
+            _host.CheckValue(_pathParser, nameof(_pathParser), "Factory failed to create a FilePathSpec");
```

Comment:
>Factory failed to create a FilePathSpec [](start = 64, length = 39)

it's not clear how FilePathSpec and factory related to each other

Reply:
I'm unsure what you're trying to say here.  I was simply following the factory pattern used for instantiating components. If there is a more appropriate way to do this, I'm all ears.

Reply:
Imagine been user who see this exception.  It can happened if PathParserFactory failed it creation, but message just say Factory (notice short name for PathParserFactory is parser) so there is no connection between them, and in general FilePathSpec doesn't provide any valuable information.

Reply:
Ahhh, I get it. Good point.